### ISSUE TITLE
Editor: store Dialog scripts outside of Game.agf

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#1008]
Among strange design decisions, Dialog scripts are not stored in separate file(s), but as a huge item group inside Game.agf (xml). If script compilers and other tools will be standalone programs, we must move them outside in their respective file or group of files.

This may be also simply convenient to user, because right now there's no way to edit them outside of editor except for searching for them and hacking XML document.

An important nuance is, that dialog scripts is a super-set of AGS script. It supports regular commands, but also has it's own. Therefore standalone script compiler won't be enough to compile them, there also has to be a preprocessor to replace special dialog commands with the proper AGS script commands. (Currently this is done by an embedded converter inside Editor program)

On another hand, maybe the final workflow could support both dialog scripts in special format and in the ags script format.

**Comment by ericoporto** [comment]
`.asd` as the suggested extension name.

**Comment by ivan-mogilko** [comment]
> .asd as the suggested extension name.

The last letter of `asc` & `ash` indicate whether they are script body or header, which may probably cause inconsistency with `asd`.

Another option is `.dsc` as in "Dialog SCript" (as opposed to ".asc" as in "Ags SCript").

**Comment by ericoporto** [comment]
Linux doesn't care for extensions, it uses magic numbers to determine a filetype primarily.

File command is the main implementation of this:
https://github.com/file/file

Note that magic numbers are not needed if the file is not binary (like .scm) and is instead a text file. If the file is a text file, then extensions are useful sugar to determine a filetype, for an external IDE that needs to know how to color a text file, like Atom when loading `.ash` and `.asc` files. But there are no limits for extension or naming of the file (like `.gitignore`). Files that start with a dot are also considered hidden by file explorers.

**Comment by morganwillcock** [comment]
I'd probably prefer `.asd` so that an editor can be configured to match with `as{c,d,h}`. Maybe the trickier choice is deciding on the file encoding.

I think this change may actually be quite complicated though, since once the content is outside of the XML file it could be modified independently of it. So unfortunately you can't really just write the text to a file and say the job is done. More likely it would be necessary to define the dialog script as something that the script editor can open to get all of the same reload behaviour and save behaviour as regular script or script header, which might mean abstracting out anything that doesn't now apply to all types of editable scripts, and anything that checks 'all scripts' (like the game writer/compiler) would need to differentiate too.

I think this is worth doing though, but whoever tries just needs to make sure that it is treated no differently to another editable text file.

**Comment by ericoporto** [comment]
I guess for now it would be the same as the `.asc` and `.ash` files uses currently - I forgot which is. I don't remember what currently blocks transition to utf-8 or other encoding and if there's an issue where this is mapped out already.

**Comment by ivan-mogilko** [comment]
TBH encoding seem to be same throughout the project files, it's ANSI (with current user system locale), and while non-unicode supportive, it is at least consistent.

**Comment by rofl0r** [comment]
> It supports regular commands, but also has it's own

you mean a separate set of script functions it can call ?

> And, of course, .ads as in "Ags Dialog Script" :).

how about simply "DialogScript.asc"
ah...

> Since dialog scripts in their original form are not exactly ags script, they should have different file extension (to be invented).

could you explain what's the difference to standard ags script ?

**Comment by ivan-mogilko** [comment]
Here's an article on dialog scripts: https://github.com/adventuregamestudio/ags-manual/wiki/Settingupthegame#conversations
you need to scroll down a little to see the list of special syntax.

In the older version of AGS dialog scripts were processed separately from main script. Later it was converted to ags script by a special preprocessor.
All of the special commands are actually replaced with ags script commands.
https://github.com/adventuregamestudio/ags/blob/ags3/Editor/AGS.Editor/Utils/DialogScriptConverter.cs

This is also why I was suggesting a separate file extension. Also, not sure about putting everything in one script, because in large games dialogs may also be very long.

**Comment by rofl0r** [comment]
ah, thanks. i think for the moment it would be sufficient if the editor saves the current converted dialog script to file with .asc extension (and maybe the original script with .asd or whatever), but in the long run it'd be nice to have a standalone command line tool that can do the conversion.

**Comment by jduo** [comment]
I can see about picking this one up.

**Comment by ivan-mogilko** [comment]
@jduo please tell what you are planning to implement when you are ready, so that we could discuss this.

**Comment by ivan-mogilko** [comment]
I wrote some points about what should be done (also adding them to the first post):

1. Storage. Now the dialog scripts are stored in Game.agf XML file, <Dialog><Script> item. Instead this script should be stored in a separate file. We need all item contents minus `<![CDATA[` opening and closing tags.
The name of the file is N.EXT, where N = Dialog's script name and EXT is special extension, yet to be decided. (But proper extension is a least concern for now, as it may be adjusted later, before new version is finalized.)

2. Saving/Loading. Right now it seems that dialog scripts are saved and loaded along with the rest of the game data, during XML parsing. This may be replaced with function for loading script files explicitly. Such function will then be called when
    * a Dialog is about to be displayed for editing;
    * Editor want to compile dialog scripts;

Saving may be done perhaps as simply an extra step when saving a game (since there's no separate save operation for dialog).

3. Hot reloading. Similar to regular scripts, which are stored as separate files already, editor should be detecting file changes to dialog script files, and reload them if they are changed, asking user if necessary. There's a editor setting that controls this behavior for normal scripts, same should be used for dialog scripts.


If someone knows other potential issues or have ideas, please ammend.

**Comment by ericoporto** [comment]
Hey, I think we can remove the label `context: script compiler` from this issue.

**Comment by persn** [comment]
@jduo Are you working on this?

**Comment by ivan-mogilko** [comment]
After consideration, I guess this belongs to "ags4" branch. This is not obligatory for building game with command-line tools, as they may easily export these dialog scripts from Game.agf.

**Comment by jduo** [comment]
> @jduo Are you working on this?

I am not working on this now.

**Comment by persn** [comment]
Okay, I'll probably take over this issue after new years

**Comment by ivan-mogilko** [comment]
Note there's a standalone tool for converting dialog scripts to regular scripts in works (#1269). For now the tool is reading dialogs from Game.agf. When this task will have some progress we might have to make another variant of the tool that would read dialogs from individual files.

**Comment by ericoporto** [comment]
I looked at this and started making a version of this and got it somewhat working... So basically the way to go here I believe it's basically copy and pasting code from Script in AGS Types to Dialog there (and adjust for differences), and picking up things from the Script Component to the Dialog Component, and then figuring out how to migrate things. The Script has a few things that assumes it's always a pair of files, so it's not currently quite easy to think on how to share it with Dialog (imagining something both inherit), but the Script Component may have things that can be picked out to some editable file utility or similar.

I also took a note on the AGS Notes plugin, but it uses a completely different approach - all file operations are in the component. 

I fully believe in making a version that works with some "repeated code" and then later figure out how to split things apart. One cool thing of this is perhaps how things are picked a part makes easier to add a generic text editor for file thing that could be borrowed for a functionality like AGS Notes that could be used in the Editor - we can pack arbitrary files nowadays so it would be cool to edit pure text files that are later used in-game as resource right in AGS editor itself.

**Comment by ericoporto** [comment]
Uhm, my main thought was to make a PlainText file object and have both Dialog and Script use that and maybe some utility methods in the Editor side to go with it.

My idea was if someone wanted to add edit capabilities to some other plain text format to the editor it could be leveraged by it.

Also, the dialog stuff that is in the xml that is not the cdata (the text file), I was thinking about just leaving in game.agf and if someone later wants to split those too, it's fine, but I don't think I want play with that part - my use-case need is to load the dialog script text file in a different editor that has advanced spell check capabilities and then I would just reload the file in the Editor.

**Comment by ivan-mogilko** [comment]
Maybe I miss something, but I don't see how what you say contradicts to my proposal.

**Comment by ericoporto** [comment]
You are right it doesn't. I should have clarified that, probably doing both.

**Comment by ericoporto** [comment]
I always thought this was super difficulty. I tried to ignore everything above and give it a shot.

https://github.com/ericoporto/ags/commit/ab5c0823015dcac552ff3433813f476449d453a7

This works apparently. The only issue is apparently `.asd` is also the name of a word recovery format (MS Word is a popular application so I expect other people to see the same icon as I)

![image](https://github.com/user-attachments/assets/fbe9cbd2-c52f-4aff-9e58-d625c9445200)

**Comment by ericoporto** [comment]
I would like to try to pick this again later, so I am rereading things.

> create a separate DialogScript class that works only with the script. Dialog will become a container class which defines dialog properties and options, and has a reference to DialogScript. This is similar to how Room class has a reference to Script object.

Ok, I understood this as create a new class then in `AGS.Types/DialogScript.cs`, and then refactor the [Dialog.Script](https://github.com/adventuregamestudio/ags/blob/5209c10679fe6e234472201fd205b8a6d31aaabb/Editor/AGS.Types/Dialog.cs#L67) property from a string to a DialogScript type, which would hold a Text property that would be similar to the previous “Script” property, would also hold a property for the file name or any rules related to it, and some way to read and write back to this file. Somehow the logic that reads from the xml should be kept so it can load old projects but we would always write to a new file so the old serialization of the content inside the Game.agf could be removed.

Everything then that used to read Dialog.Script would now read Dialog.Script.Text.

**Comment by ivan-mogilko** [comment]
> Somehow the logic that reads from the xml should be kept so it can load old projects but we would always write to a new file so the old serialization of the content inside the Game.agf could be removed.

Yes, this logic should stay in Dialog's `Dialog(XmlNode node)` constructor, but assume that the script data may be missing.

**Comment by ericoporto** [comment]
I know this is meant to ags4, but I will do some initial work in master because I find it easier to work there, I just happen to have more of my own game to throw to test things...

Here's a first commit https://github.com/ericoporto/ags/commit/d34ff80fd53f444bd22a2bb7104d88b93449406d

I thought about having the dialog files in a Dialogs directory, similar to ags4 (but without the ID subdirs), but was surprised the `Room.cs` doesn't define the directory the rooms are, so not sure where to put this information. I thought about adding a function like this in DialogScript.

```C#
public static string GetFileName(Dialog dialog)
{
    return Path.Combine("Dialogs", dialog.Name + ".asd");
}
```

**Comment by ivan-mogilko** [comment]
> I know this is meant to ags4, but I will do some initial work in master because I find it easier to work there, I just happen to have more of my own game to throw to test things...

Do you mean that you want to put this feature into the 3.x branch, or that you will experiment in master branch and then move the feature to ags4?

3.x version should not have any major changes to project structure, and the switch to ags4 is long overdue (mostly because of me). I'm currently planning to wrap the 3.6.3, adding few last things.

Also, ags4 now has a different way to run a project upgrade, through the number of "upgrade tasks" that may be displayed as pages in the Upgrade Game wizard dialog.

**Comment by ivan-mogilko** [comment]
> I thought about having the dialog files in a Dialogs directory, similar to ags4 (but without the ID subdirs), but was surprised the Room.cs doesn't define the directory the rooms are

The directory is defined in UnloadedRoom class. This is the parent class for the Room, which represents a project item, without much of the actual data.

**Comment by ericoporto** [comment]
Oh, yeah, the idea is to sketch on master and then rebase to ags4 to work on the details, specifically the filewatcher stuff is one thing I would only work after rebasing.

**Comment by ericoporto** [comment]
@ivan-mogilko Hey! This branch is actually working: https://github.com/ericoporto/ags/tree/feature-external-dialog .

I redid a version over ags4: https://github.com/ericoporto/ags/tree/ags4--feature-external-dialog (simple cherry-pick worked!)

But now I need help... With two things!

So, on ags4 you mentioned there is a new Upgrade Game wizard dialog. I don't know how to use it, that is one problem. (I read the PR in https://github.com/adventuregamestudio/ags/pull/2794, but I am curious, should the object Dialog and DialogScript be modified to only read and write the actual files and the reading from the previous xml CDATA be made into a conversion task to be applied by the upgrade dialog?)

The other issue I need help is with the file watcher, I noticed the file watch stuff triggers in the ScriptEditor, and there isn't one in the DialogEditor. Would it be better to move it to ScriptEditorBase or to duplicate in the DialogEditor?

I guess I can put a draft PR for it and discuss there, while I can work on understanding a few things I don’t understand (I put a lot of comments on things I was in doubts)

**Comment by ivan-mogilko** [comment]
Indeed, it's best to open a PR draft.

> So, on ags4 you mentioned there is a new Upgrade Game wizard dialog. I don't know how to use it, that is one problem. (I read the PR in https://github.com/adventuregamestudio/ags/pull/2794, but I am curious, should the object Dialog and DialogScript be modified to only read and write the actual files and the reading from the previous xml CDATA be made into a conversion task to be applied by the upgrade dialog?)

I think that old data should be read and converted by a upgrade task. This is a one-time operation, and modern Dialog class should not know anything about old format.
Upgrade tasks are found in project here:
https://github.com/adventuregamestudio/ags/tree/ags4/Editor/AGS.Editor/Entities/UpgradeGame
each is represented by a class.

The [IUpgradeGameTask](https://github.com/adventuregamestudio/ags/blob/ags4/Editor/AGS.Editor/Entities/UpgradeGame/IUpgradeGameTask.cs) is a interface that must be implemented by every upgrade task class.
The task class may act silently, or provide a Wizard Page with a text, and some options (if necessary). Dialog conversion is probably similar to Room conversion or Fonts conversion, it is done unconditionally, so no option controls are required, only a page with text explaining what kind of changes is going to be performed. See [UpgradeGameFontsToFontFilesTask](https://github.com/adventuregamestudio/ags/blob/ags4/Editor/AGS.Editor/Entities/UpgradeGame/UpgradeGameFontsToFontFilesTask.cs) and [UpgradeGameRoomsOpenFormatTask](https://github.com/adventuregamestudio/ags/blob/ags4/Editor/AGS.Editor/Entities/UpgradeGame/UpgradeGameRoomsOpenFormatTask.cs) for example.

For adding upgrade tasks there's GamePrepareUpgrade event:
https://github.com/adventuregamestudio/ags/blob/fed0779bd74dc0675d8dc515139bf1ba8842b8c3/Editor/AGS.Editor/EditorEvents.cs#L16-L17

In DialogsComponent subscribe for this event, and add respective task in the handler, similar to this:
https://github.com/adventuregamestudio/ags/blob/fed0779bd74dc0675d8dc515139bf1ba8842b8c3/Editor/AGS.Editor/Components/FontsComponent.cs#L530-L533

**Comment by ericoporto** [comment]
ok, I started a draft PR in https://github.com/adventuregamestudio/ags/pull/2978 , I will later add a description to it and continue working on it

- still need to do proper handling of renaming
- need to finish making the upgrade task
- need to figure out what to do about file watcher and script editor or dialog editor
